scvid/module/incremental_pca.py

```diff
@@ -0,0 +1,119 @@
+# Copyright Contributors to the Cellarium project.
+# SPDX-License-Identifier: BSD-3-Clause
+
+import math
+
+import torch
+import torch.nn as nn
+
+from scvid.module import BaseModule
+
+
+class IncrementalPCA(BaseModule):
+    """
+    Incremental PCA.
+
+    **Reference:**
+
+    1. *A Distributed and Incremental SVD Algorithm for Agglomerative Data Analysis on Large Networks*,
+       M. A. Iwen, B. W. Ong
+       (https://users.math.msu.edu/users/iwenmark/Papers/distrib_inc_svd.pdf)
+    2. *Incremental Learning for Robust Visual Tracking*,
+       D. Ross, J. Lim, R.-S. Lin, M.-H. Yang
+       (https://www.cs.toronto.edu/~dross/ivt/RossLimLinYang_ijcv.pdf)
+
+    Args:
+        g_genes: Number of genes.
+        k_components: Number of principal components.
+        transform: If not ``None`` is used to transform the input data.
+        perform_mean_correction: If ``True`` then the mean correction is applied to the update step.
+            If ``False`` then the data is assumed to be centered and the mean correction
+            is not applied to the update step.
+    """
+
+    def __init__(
+        self,
+        g_genes: int,
+        k_components: int,
+        perform_mean_correction: bool = False,
+        transform: nn.Module | None = None,
+    ) -> None:
+        super().__init__()
+        self.g_genes = g_genes
+        self.k_components = k_components
+        self.perform_mean_correction = perform_mean_correction
+        self.transform = transform
+        self.V_kg: torch.Tensor
+        self.S_k: torch.Tensor
+        self.x_mean_g: torch.Tensor
+        self.x_size: torch.Tensor
+        self.register_buffer("V_kg", torch.zeros(k_components, g_genes))
+        self.register_buffer("S_k", torch.zeros(k_components))
+        self.register_buffer("x_mean_g", torch.zeros(g_genes))
+        self.register_buffer("x_size", torch.tensor(0))
+        self._dummy_param = torch.nn.Parameter(torch.tensor(0.0))
+
+    @staticmethod
+    def _get_fn_args_from_batch(
+        tensor_dict: dict[str, torch.Tensor]
+    ) -> tuple[tuple, dict]:
+        x = tensor_dict["X"]
+        return (x,), {}
+
+    def forward(self, x_ng: torch.Tensor) -> None:
+        if self.transform is not None:
+            x_ng = self.transform(x_ng)
+
+        g = self.g_genes
+        k = self.k_components
+        m = self.x_size
+        n = x_ng.size(0)
+        assert k <= min(n, g), (
+            f"Rank of svd_lowrank (k_components): {k}"
+            f" must be less than min(n_cells, g_genes): {min(n, g)}"
+        )
+
+        # compute SVD of new data
+        if self.perform_mean_correction:
+            x_mean_g = x_ng.mean(dim=0)
+            _, S_k, V_gk = torch.svd_lowrank(x_ng - x_mean_g, q=k)
```

Comment:
Please expose `niter` and make it kwarg of `IncrementalPCA` and call it `svd_lowrank_niter`.  Also, expose it to CLI.  No need to test it.

Reply:
Done!